### PR TITLE
Disable prompts to save logins while on extension pages

### LINKED ISSUE #57: Disable login prompts in Management Interface

### PR DESCRIPTION
**Pull request by lmorchard**
- Add getLoginSavingEnabled and setLoginSavingEnabled methods to the
  Logins experiment API

- Disable login saving for extension pages on background startup

Fixes #57

### CODE REVIEW

File: test/integration/logins-api-test.js [hunk 1]
```diff
@@ -235,6 +249,45 @@ describe("logins API", () => {
...
+      it("enables login saving for origin when given true", async () => {
+        await commonTest(true, true);
+      });
+    });
+    describe("with pref signon.rememberSignons = false", () => {
+      before(async () => {
+        await disableRememberSignons();
+      });
+      it("does not enable login saving for origin when given true", async () => {
+        await commonTest(true, false);
```
Comment (lmorchard) [inline test/integration/logins-api-test.js]: Mild corner case here: If the user has turned off saving logins globally in options, `getLoginSavingEnabled` always returns false no matter what the setting might be.

File: src/background/environment.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+export async function initializeEnvironment() {
+  // Issue #57: Disable login saving for extension pages.
+  const origin = browser.extension.getURL("").slice(0, -1);
+  const savingEnabled = await browser.experiments.logins
+    .getLoginSavingEnabled(origin);
+  if (savingEnabled) {
```
Comment (lmorchard) [inline src/background/environment.js]: Checking whether saving is enabled to ensure we keep it off. The user could enable login save globally or manually delete the exception from the list in options. Kind of edge casey, but bleh.

**Comment by lmorchard** [comment]
Hmm, so I only ensure saving is off on startup of the addon. The user could later enable saving in options or delete the exception - and then we'd see the login save prompt come up until next add-on startup. 

I wonder if I should make a check more often? Like, say, whenever the browser action button is clicked?

**Comment by linuxwolf** [comment]
> Hmm, so I only ensure saving is off on startup of the addon. The user could later enable saving in options or delete the exception - and then we'd see the login save prompt come up until next add-on startup.
> 
> I wonder if I should make a check more often? Like, say, whenever the browser action button is clicked?

I think we can start with this, and see how it goes.  If we do it more frequently, I'd argue for whenever we think the management page is loaded.

File: src/background/environment.js (hunk 2 above)
Comment (6a68) [reply src/background/environment.js]: It seems unlikely that users are going to mess with this manually. That said, the browser APIs are available from extension pages, so we could just run this check when the management page is loaded (or, to get kind of extreme about the "user messes things up" edge case, we could run this check on page load and whenever the page visibility API's `visibilitychange` event fires).

File: src/experiments/logins/api.js [hunk 3]
```diff
@@ -49,6 +49,20 @@ this.logins = class extends ExtensionAPI {
           //   - fooLogin = login as vanilla JS object
           //   - fooLoginInfo = login as nsILoginInfo
           //   - fooBag = (subset of) login fields as nsIPropertyBag
+          getLoginSavingEnabled(aHost) {
```
Comment (6a68) [inline src/experiments/logins/api.js]: I really dislike the `aFoo` naming convention. How about just calling this `origin`?

File: src/experiments/logins/api.js [hunk 4]
```diff
@@ -49,6 +49,20 @@ this.logins = class extends ExtensionAPI {
...
+            try {
+              return Services.logins.getLoginSavingEnabled(aHost);
```
Comment (6a68) [inline src/experiments/logins/api.js]: The trailing slash thing seems like a super annoying gotcha. I think we should either handle it here (strip it out or throw if it's detected), or document this annoyance in the schema docs.

File: src/experiments/logins/schema.json [hunk 5]
```diff
@@ -199,6 +199,32 @@
       }
     ],
     "functions": [
+      {
+        "name": "getLoginSavingEnabled",
+        "type": "function",
+        "description": "For the given origin, gets a boolean flag which determines whether login saving is enabled",
```
Comment (6a68) [inline src/experiments/logins/schema.json]: Should also mention that the returned promise is rejected if the URL is malformed.

File: src/experiments/logins/schema.json [hunk 6]
```diff
@@ -199,6 +199,32 @@
...
+        "async": true,
+        "parameters": [{
+          "name": "aHost",
+          "type": "string",
+          "description": "The origin for which to check the status of login saving"
```
Comment (6a68) [inline src/experiments/logins/schema.json]: Just use the description from [the IDL](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsILoginManager#getLoginSavingEnabled()): `The hostname to check. This argument should be in the origin URL format, with no pathname. For example: "https://www.site.com".`

File: src/experiments/logins/schema.json [hunk 7]
```diff
@@ -199,6 +199,32 @@
...
+        }]
+      },
+      {
+        "name": "setLoginSavingEnabled",
+        "type": "function",
+        "description": "For the given origin, uses the given value to set a boolean flag which determines whether login saving is enabled",
```
Comment (6a68) [inline src/experiments/logins/schema.json]: The IDL docs may be clearer here. Also, be sure to mention cases in which the returned promise is rejected, which I think is just if the URL is malformed.

File: src/experiments/logins/api.js [hunk 8]
```diff
@@ -49,6 +49,20 @@ this.logins = class extends ExtensionAPI {
...
+            } catch (ex) {
+              throw new ExtensionError(ex);
+            }
+          },
+          setLoginSavingEnabled(aHost, isEnabled) {
```
Comment (6a68) [inline src/experiments/logins/api.js]: How about `origin` and `shouldSaveLogins` for the arguments?

File: test/integration/logins-api-test.js [hunk 9]
```diff
@@ -46,6 +46,20 @@ describe("logins API", () => {
   let By;
   let until;
 
+  const enableRememberSignons = async () => {
+    await webext.inChrome();
+    await driver.executeScript(`
+      Services.prefs.setBoolPref("signon.rememberSignons", true);
```
Comment (6a68) [inline test/integration/logins-api-test.js]: It would be good to actually call the Services.logins.get/setLoginSavingEnabled for a specific origin, then test it's correctly returned when you call the API. I wonder what happens if you accidentally leave in a trailing slash, for instance. This could be left for a followup commit.

File: test/integration/logins-api-test.js (hunk 1 above)
Comment (6a68) [reply test/integration/logins-api-test.js]: Good catch! I'm not sure which, if any, parts of the logins API will work if logins are globally disabled. I think this needs investigation; could you file a bug?

File: src/experiments/logins/schema.json (hunk 6 above)
Comment (6a68) [reply src/experiments/logins/schema.json]: Actually, looking at the schema docs, I haven't been super strict about reusing the IDL descriptions. This could be fine as is (but `aHost` does bug me ;-P).

**Comment by 6a68** [comment]
Whoops! Didn't see Matt had already reviewed. Sorry to pile on :-P

File: test/integration/logins-api-test.js (hunk 9 above)
Comment (linuxwolf) [reply test/integration/logins-api-test.js]: From my own experiences, the CRUD portion of the upstream API works just fine regardless of what the global pref and this specific callout are set to.

File: test/integration/logins-api-test.js (hunk 1 above)
Comment (linuxwolf) [reply test/integration/logins-api-test.js]: From my own experiences, the CRUD portion of the upstream API works just fine regardless of what the global pref and this specific callout are set to.

**Comment by lmorchard** [comment]
I added a commit that automatically re-adds our entry to the exception list for login saving when it is somehow removed (manually or otherwise). However, this still needs unit / integration tests. However, however, this could use a sanity check as to whether this is what we actually want to do.

File: src/experiments/logins/api.js (hunk 3 above)
Comment (lmorchard) [reply src/experiments/logins/api.js]: Heh, I used aHost to match the upstream API, but I can change it if necessary

File: src/background/environment.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  // If login saving is enabled for us on startup, add an exception.
+  const savingEnabled = await browser.experiments.logins
+    .getLoginSavingEnabled(ourOrigin);
+  if (savingEnabled) {
+    await browser.experiments.logins
+      .setLoginSavingEnabled(ourOrigin, false);
+  }
+
+  // HACK: If our exception to login saving is removed, immediately re-add it.
+  browser.experiments.logins.onLoginSavingChanged
```
Comment (lmorchard) [inline src/background/environment.js]: Okay, here's the bit that could use some sanity checking. I worked out how to subscribe to the event that fires when a login saving exception is removed, and automatically re-add it. This may be a mistake, but I can't think of a reason why

File: src/experiments/logins/api.js (hunk 8 above)
Comment (lmorchard) [reply src/experiments/logins/api.js]: Yeah, same as above, I was trying to follow the underlying API rather than invent new names

**Comment by lmorchard** [comment]
Actually, you know what: Never mind. It occurred to me over the weekend that this PR *without* the event handler to automatically re-enable the exception was already 80% of the solution and pretty much reviewed.

My commit that uses the permission events still needs time to bake, so I'll do it in a follow-up PR.

**Comment by punkbusterdani** [comment]
I wish to verify the fix of this issue, but I'm lacking some steps here. How am I supposed to reproduce/verify this issue? Thanks.
